tpm2/platform.go

```diff
@@ -111,11 +144,115 @@ func (h *platformKeyDataHandler) RecoverKeys(data *secboot.PlatformKeyData) (sec
+	}
+	defer tpm.Close()
+
+	var k *SealedKeyData
+	if err := json.Unmarshal(handle, &k); err != nil {
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  err}
+	}
+	if k.data.Version() < 3 {
+		// All KeyData objects created for this platform are at least v3, so this
+		// should never fail. Test it though to avoid a panic later on if the version
+		// has been manipulated.
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  fmt.Errorf("invalid key data version: %d", k.data.Version())}
+	}
+
+	// Validate the initial key data
+	_, err = k.validateData(tpm.TPMContext, tpm.HmacSession())
+	switch {
+	case isKeyDataError(err):
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  err}
+	case err != nil:
+		return nil, xerrors.Errorf("cannot validate key data: %w", err)
+	}
+
+	srk, err := tpm.CreateResourceContextFromTPM(tcg.SRKHandle)
+	switch {
+	case tpm2.IsResourceUnavailableError(err, tcg.SRKHandle):
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorUninitialized,
+			Err:  ErrTPMProvisioning}
+	case err != nil:
+		return nil, xerrors.Errorf("cannot create context for SRK: %w", err)
+	}
+
+	keyObject, err := k.load(tpm.TPMContext, srk, tpm.HmacSession())
+	switch {
+	case isLoadInvalidParamError(err) || isImportInvalidParamError(err):
+		// The supplied key data is invalid or is not protected by the supplied SRK.
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  xerrors.Errorf("cannot load sealed key object into TPM: %w", err)}
+	case isLoadInvalidParentError(err) || isImportInvalidParentError(err):
+		// The supplied SRK is not a valid storage parent.
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorUninitialized,
+			Err:  ErrTPMProvisioning}
+	case err != nil:
+		// This is an unexpected error
+		return nil, xerrors.Errorf("cannot load sealed key object into TPM: %w", err)
+	}
+	defer tpm.FlushContext(keyObject)
+
+	if len(old) > 0 {
+		v, err := deriveAuthValue(old, keyObject.Name().Algorithm().Size())
+		if err != nil {
+			return nil, xerrors.Errorf("cannot derive old auth value: %w", err)
+		}
+		keyObject.SetAuthValue(v)
+	}
+
+	var newAuthValue []byte
+	if len(new) > 0 {
+		newAuthValue, err = deriveAuthValue(new, keyObject.Name().Algorithm().Size())
+		if err != nil {
+			return nil, xerrors.Errorf("cannot derive new auth value: %w", err)
+		}
+	}
+
+	priv, err := tpm.ObjectChangeAuth(keyObject, srk, newAuthValue, tpm.HmacSession().IncludeAttrs(tpm2.AttrCommandEncrypt))
+	if err != nil {
+		if tpm2.IsTPMSessionError(err, tpm2.ErrorAuthFail, tpm2.CommandObjectChangeAuth, 1) {
+			return nil, &secboot.PlatformHandlerError{
+				Type: secboot.PlatformHandlerErrorInvalidAuthKey,
+				Err:  err}
+		}
```

Comment:
shouldn't the case where the wrong old key is provided be tested?, where will that fail? here?

Reply:
I'll add some extra tests for this.